cdap-app-fabric/src/main/java/io/cdap/cdap/internal/credential/store/SecureMultiStore.java

```diff
@@ -0,0 +1,267 @@
+import io.cdap.cdap.api.NamespaceResourceManager;
+import io.cdap.cdap.api.NamespaceResourceNotFoundException;
+import io.cdap.cdap.api.NamespaceResourceReference;
+import io.cdap.cdap.api.common.Bytes;
+import io.cdap.cdap.api.security.AccessException;
+import io.cdap.cdap.api.security.credential.Secured;
+import io.cdap.cdap.api.security.store.SecureStore;
+import io.cdap.cdap.api.security.store.SecureStoreManager;
+import io.cdap.cdap.proto.id.NamespaceId;
+import io.cdap.cdap.spi.data.StructuredTable;
+import io.cdap.cdap.spi.data.StructuredTableContext;
+import io.cdap.cdap.spi.data.table.field.Field;
+import io.cdap.cdap.spi.data.table.field.Fields;
+import io.cdap.cdap.spi.data.transaction.TransactionRunner;
+import io.cdap.cdap.spi.data.transaction.TransactionRunners;
+import io.cdap.cdap.store.StoreDefinition;
+import java.io.IOException;
+import java.nio.charset.StandardCharsets;
+import java.security.MessageDigest;
+import java.security.NoSuchAlgorithmException;
+import java.util.ArrayList;
+import java.util.Collection;
+import java.util.Collections;
+import java.util.ConcurrentModificationException;
+import java.util.Optional;
+import java.util.UUID;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+/**
+ * Helps maintain data consistency for a service using secure store in addition to structured
+ * tables.
+ *
+ * @param <T> The metadata class for the stored entity.
+ * @param <R> The entity class from which secured data may be extracted.
+ */
+public abstract class SecureMultiStore<T, R extends Secured<T>>
+    implements NamespaceResourceManager<R> {
+
+  private static final Logger LOG = LoggerFactory.getLogger(SecureMultiStore.class);
+
+  final TransactionRunner transactionRunner;
+
+  private final SecureStore secureStore;
+  private final SecureStoreManager secureStoreManager;
+
+  public SecureMultiStore(SecureStore secureStore, SecureStoreManager secureStoreManager,
+      TransactionRunner transactionRunner) {
+    this.secureStore = secureStore;
+    this.secureStoreManager = secureStoreManager;
+    this.transactionRunner = transactionRunner;
+  }
+
+  /**
+   * An identifier that uniquely identifies the class which utilizes secure store.
+   */
+  abstract String getResourceType();
+
+  abstract R constructFromSecureData(T metadata, String secureData);
+
+  abstract Optional<VersionedObject<T>> getMetadataInternal(StructuredTableContext context,
+      NamespaceResourceReference ref) throws IOException;
+
+  abstract Collection<NamespaceResourceReference> listMetadataInternal(
+      StructuredTableContext context, String namespace) throws IOException;
+
+  abstract void writeMetadataInternal(StructuredTableContext context,
+      NamespaceResourceReference ref, T obj, String version) throws IOException;
+
+  abstract void deleteMetadataInternal(StructuredTableContext context,
+      NamespaceResourceReference ref) throws IOException, IllegalStateException;
+
+  class VersionedObject<X> {
```

Comment:
do we need versions?

Reply:
Since we're moving away from using SecureStore, this is no longer needed.